### PR TITLE
Feature/cdap 2733 add TPFS as a source for adapters

### PR DESCRIPTION
**Pull request by rohanahata**
https://issues.cask.co/browse/CDAP-2733
http://builds.cask.co/browse/CDAP-DUT2071
This pull request aims to add TimePartitionedFileSets as an Source for Adapters.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    AvroJob.setInputKeySchema(job, avroSchema);
+  }
+
+  @Override
+  public void transform(KeyValue<AvroKey<GenericRecord>, NullWritable> input,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    System.err.println("DEBUGGING - " + input.getKey().datum().getSchema().toString());
+    emitter.emit(recordTransformer.transform(input.getKey().datum()));
+  }
+
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java]: quite a few empty newlines

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    context.setInput(tpfsAvroSourceConfig.name, source);
+    org.apache.avro.Schema avroSchema = new org.apache.avro.Schema.Parser().parse(tpfsAvroSourceConfig.schema);
+    Job job = context.getHadoopJob();
+    AvroJob.setInputKeySchema(job, avroSchema);
+  }
+
+  @Override
+  public void transform(KeyValue<AvroKey<GenericRecord>, NullWritable> input,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    System.err.println("DEBUGGING - " + input.getKey().datum().getSchema().toString());
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java]: remove?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 3]
```diff
@@ -83,8 +90,28 @@ public void testStreamConversion() throws Exception {
     DataSetManager<TimePartitionedFileSet> fileSetManager = getDataset(filesetName);
     TimePartitionedFileSet fileSet = fileSetManager.get();
 
-    List<GenericRecord> records = readOutput(fileSet, EVENT_SCHEMA);
+    List<GenericRecord> records = readOutput(fileSet, EVENT_SCHEMA_1);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: add a comment saying why you are using the new schema.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 4]
```diff
@@ -63,6 +63,13 @@
     Schema.Field.of("num", Schema.of(Schema.Type.INT)),
     Schema.Field.of("price", Schema.of(Schema.Type.DOUBLE)));
 
+  private static final Schema EVENT_SCHEMA_1 = Schema.recordOf(
+    "streamEvent",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("ticker", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("num", Schema.of(Schema.Type.INT)),
+    Schema.Field.of("price", Schema.of(Schema.Type.DOUBLE)));
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: add a comment explaining why you have dropped the field header

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 5]
```diff
@@ -83,8 +90,28 @@ public void testStreamConversion() throws Exception {
...
+
+    newManager.start();
+    newManager.waitForOneRunToFinish(4, TimeUnit.MINUTES);
+    newManager.stop();
+
+    DataSetManager<TimePartitionedFileSet> newFileSetManager = getDataset(newFilesetName);
+    TimePartitionedFileSet newFileSet = newFileSetManager.get();
+
+    List<GenericRecord> newRecords = readOutput(newFileSet, EVENT_SCHEMA_1);
+    Assert.assertEquals(1, newRecords.size());
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: extract this into a method ?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 6]
```diff
@@ -83,8 +90,28 @@ public void testStreamConversion() throws Exception {
...
     Assert.assertEquals(1, records.size());
+
+
+    String newFilesetName = filesetName + "_op";
+    ETLBatchConfig etlBatchConfig = constructTPFSETLConfig(filesetName, newFilesetName);
+
+    AdapterConfig newAdapterConfig = new AdapterConfig("description", TEMPLATE_ID.getId(),
+                                                       GSON.toJsonTree(etlBatchConfig));
+    Id.Adapter newAdapterId = Id.Adapter.from(NAMESPACE, "sconversion1");
+    AdapterManager newManager = createAdapter(newAdapterId, newAdapterConfig);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: rename it , tpfsAdapterManager?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+/**
+ * A {@link BatchSource} to read Avro record from {@link TimePartitionedFileSet}
+ */
+@Plugin(type = "source")
+@Name("TPFSAvro")
+@Description("AVRO Source with Time Partitioned File Dataset")
+public class TimePartitionedFileSetDatasetAvroSource extends
+  BatchSource<AvroKey<GenericRecord>, NullWritable, StructuredRecord> {
+
+  private static final String SCHEMA_DESC = "The avro schema of the record being written to the Sink as a JSON Object";
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java]: source, not sink. fix description.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+  private static final String TPFS_NAME_DESC = "Name of the Time Partitioned FileSet Dataset to which the records " +
+    "have to be written. If it doesn't exist, it will be created";
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java]: same here

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/AvroToStructuredTransformer.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import com.google.common.collect.Maps;
+import org.apache.avro.generic.GenericRecord;
+import java.util.Map;
+
+/**
+ *
+ */
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/AvroToStructuredTransformer.java]: add javadoc

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+  private static final String SCHEMA_DESC = "The avro schema of the record being read from the Source as a JSON Object";
+  private static final String TPFS_NAME_DESC = "Name of the Time Partitioned FileSet Dataset to which the records " +
+    "have to be read from.";
+  private static final String BASE_PATH_DESC = "Base path for the time partitioned fileset. Defaults to the " +
+    "name of the dataset";
+  private static final String DURATION_DESC = "Size of the time window to read with each run of the pipeline. " +
+    "The format is expected to be a number followed by a 's', 'm', 'h', or 'd' specifying the time unit, with 's' " +
+    "for seconds, 'm' for minutes, 'h' for hours, and 'd' for days. For example, a value of '5m' means each run of " +
+    "the pipeline will read 5 minutes of events from the stream.";
+  private static final String DELAY_DESC = "Optional delay for reading stream events. The value must be " +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java]: description talks about stream instead of tpfs

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    @Description(BASE_PATH_DESC)
+    @Nullable
+    private String basePath;
+
+    @Description(DURATION_DESC)
+    private String duration;
+
+    @Description(DELAY_DESC)
+    @javax.annotation.Nullable
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/TimePartitionedFileSetDatasetAvroSource.java]: use the javax annotation everywhere instead of the avro one

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/AvroToStructuredTransformer.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+    if (schemaCache.containsKey(hashCode)) {
+      structuredSchema = schemaCache.get(hashCode);
+    } else {
+      structuredSchema = Schema.parseJson(genericRecordSchema.toString());
+      schemaCache.put(hashCode, structuredSchema);
+    }
+
+    StructuredRecord.Builder builder = StructuredRecord.builder(structuredSchema);
+    for (Schema.Field field: structuredSchema.getFields()) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/AvroToStructuredTransformer.java]: this will not work for more complex schemas, such as when a record has a field that is a record.  Or if Avro is using its utf8 class to represent strings.  This is also true of the structured to avro transformer this is based off of.

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/AvroToStructuredTransformer.java]: I would do this part in a separate PR and merge that before merging this PR.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 13]
```diff
@@ -77,6 +86,49 @@ public void testStreamConversionTPFSAvroSink() throws Exception {
...
+
+    manager.start();
+    manager.waitForOneRunToFinish(4, TimeUnit.MINUTES);
+    manager.stop();
+
+    // get the output fileset, and read the parquet/avro files it output.
+    DataSetManager<TimePartitionedFileSet> fileSetManager = getDataset(filesetName);
+    TimePartitionedFileSet fileSet = fileSetManager.get();
+
+    List<GenericRecord> records = readOutput(fileSet, EVENT_SCHEMA_1);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: looks like the first part of this test is only used to create an avro file. This is unnecessary and adds a couple minutes to the test. You can just create the input fileset in the test case directly, then write the input you want to write so that there is just a single mapreduce run in the test.

**Comment by albertshau** [comment]
one comment about test efficiency

**Comment by albertshau** [comment]
if you want, can do it in a separate pr.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 14]
```diff
@@ -65,6 +65,15 @@
     Schema.Field.of("num", Schema.of(Schema.Type.INT)),
     Schema.Field.of("price", Schema.of(Schema.Type.DOUBLE)));
 
+  // The new event scheme is being used over the old schema because Maps when parsed take ordering of keys into account.
+  // Once CDAP-2813 is resolved, we can switch to using EVENT_SCHEMA
+  private static final Schema EVENT_SCHEMA_1 = Schema.recordOf(
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: Give it a better name. Maybe EVENT_SCHEMA_WITHOUT_HEADER

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java [hunk 15]
```diff
@@ -77,6 +86,49 @@ public void testStreamConversionTPFSAvroSink() throws Exception {
...
+    tpfsAdapterManager.start();
+    tpfsAdapterManager.waitForOneRunToFinish(4, TimeUnit.MINUTES);
+    tpfsAdapterManager.stop();
+
+    DataSetManager<TimePartitionedFileSet> newFileSetManager = getDataset(newFilesetName);
+    TimePartitionedFileSet newFileSet = newFileSetManager.get();
+
+    List<GenericRecord> newRecords = readOutput(newFileSet, EVENT_SCHEMA_1);
+    Assert.assertEquals(1, newRecords.size());
+
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLStreamConversionTest.java]: redundant newline

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+  private static final Gson GSON = new Gson();
+  private static Schema EVENT_SCHEMA_WITHOUT_HEADER;
+  @Test
+  public void testAvroSourceConversionToAvroSink() throws Exception {
+
+    Schema schema = Schema.recordOf(
+      "record",
+      Schema.Field.of("int", Schema.of(Schema.Type.INT)));
+    EVENT_SCHEMA_WITHOUT_HEADER = schema;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java]: its not good to pass things around by setting static variables, since anybody can access them. If you need to pass something, pass it as an argument of the method.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    tpfsAdapterManager.stop();
+
+    DataSetManager<TimePartitionedFileSet> newFileSetManager = getDataset(newFilesetName);
+    TimePartitionedFileSet newFileSet = newFileSetManager.get();
+
+    List<GenericRecord> newRecords = readOutput(newFileSet, EVENT_SCHEMA_WITHOUT_HEADER);
+    Assert.assertEquals(1, newRecords.size());
+  }
+
+  private org.apache.avro.Schema convertSchema(Schema cdapSchema) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java]: don't need a separate method for this, it's just in one place. Just call the line directly.

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java]: it's also weird to convert it. Everywhere else in this class, the avro schema is used. You should be able to just create it directly.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+      .build());
+    DataSetManager<TimePartitionedFileSet> fileSetManager = getDataset(filesetName);
+    TimePartitionedFileSet tpfs = fileSetManager.get();
+
+    TransactionManager txService = getTxService();
+    Transaction tx1 = txService.startShort(100);
+    TransactionAware txTpfs = (TransactionAware) tpfs;
+    txTpfs.startTx(tx1);
+
+    fileSetManager.get().addPartition(System.currentTimeMillis(), "directory", ImmutableMap.of("key1", "value1"));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java]: its possible, though unlikely, that the time on line and the next line will cross a partition boundary. For example, the minute becomes different.  It is better to get the time once, then use that value in both of these calls.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+    tpfsAdapterManager.start();
+    tpfsAdapterManager.waitForOneRunToFinish(4, TimeUnit.MINUTES);
+    tpfsAdapterManager.stop();
+
+    DataSetManager<TimePartitionedFileSet> newFileSetManager = getDataset(newFilesetName);
+    TimePartitionedFileSet newFileSet = newFileSetManager.get();
+
+    List<GenericRecord> newRecords = readOutput(newFileSet, EVENT_SCHEMA_WITHOUT_HEADER);
+    Assert.assertEquals(1, newRecords.size());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java]: it would be better to validate the record value as well

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java (hunk 17 above)
Comment (rohanahata) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLTPFSTest.java]: we need to  convert it because the eventSchemaWithoutHeader (which in this commit has been changed to eventSchema) takes in a Schema of StructuredRecords

**Comment by albertshau** [comment]
lgtm
